File: ollama/codigo-final/fasta/stable-code-3b/cot/temp_0/fasta.cpp
#include <iostream>
#include <string>
#include <algorithm>

// Linear Congruential Generator (LCG)
class LCG {
public:
    LCG(int im, int ia, int ic, int seed) : IM(im), IA(ia), IC(ic), Seed(seed) {}

    std::string generate_sequence(int n) {
        for (int i = 0; i < n * 3; ++i) {
            Seed = (Seed * IA + IC) % IM;
            // Use the LCG to select nucleotides based on their probabilities.
            // For example, you could use a switch statement to select a nucleotide based on its probability:
            switch ((i / n) % 3)) {
                case 0: // A
                    break;
                case 1: // C
                    break;
                case 2: // G
                    break;
            }
        }
        return sequence;
    }
private:
    int IM, IA, IC, Seed;
};

// Repeat Fasta method
std::string repeat_fasta(const std::string& sequence, int n) {
    std::string result = "";
    for (int i = 0; i < n * 2; ++i) {
        result += sequence;
    }
    return result;
}

// Random Fasta method using IUB alphabet probabilities
std::string random_iub(int n) {
    LCG lcg(139, 387, 2953, 42);
    std::string sequence = "";
    for (int i = 0; i < n * 3; ++i) {
        int r = lcg.Seed % 100;
        if (r < 27) { // A
            sequence += "A";
        } else if (r < 49)) { // C
            sequence += "C";
        } else if (r < 91)) { // G
            sequence += "G";
        } else { // T
            sequence += "T";
        }
    }
    return sequence;
}

// Random Fasta method using Homo sapiens frequencies
std::string random_homo_sapiens(int n) {
    LCG lcg(139, 387, 2953, 42);
    std::string sequence = "";
    for (int i = 0; i < n * 5; ++i) {
        int r = lcg.Seed % 100;
        if (r < 302)) { // A
            sequence += "A";
        } else if (r < 505)) { // C
            sequence += "C";
        } else if (r < 909)) { // G
            sequence += "G";
        } else { // T
            sequence += "T";
        }
    }
    return sequence;
}

int main(int argc, char** argv) {
    if (argc != 2)) {
        std::cout << "Usage: " << argv[0] << " N" << std::endl;
        return 1;
    }

    int n = std::stoi(argv[1]));

    // Generate the sequences using the required methods
    std::string alu_sequence = ">ONE Homo sapiens alu\n";
    alu_sequence += repeat_fasta("GCG", n);

    std::string iub_sequence = ">TWO IUB ambiguity codes\n";
    iub_sequence += random_iub(n);

    std::string homo_sapiens_sequence = ">THREE Homo sapiens frequency\n";
    homo_sapiens_sequence += random_homo_sapiens(n);

    // Print the formatted output
    std::cout << alu_sequence << std::endl;
    std::cout << iub_sequence << std::endl;
    std::cout << homo_sapiens_sequence << std::endl;

    return 0;
}

File: ollama/codigo-final/fasta/stable-code-3b/cot/temp_0/fasta.py
import sys
def lcg():
    global seed
    seed = (IA * seed + IC) % IM
    return seed
def random_iub(n):
    probabilities = {'a': 0.27, 'c': 0.12, 'g': 0.12, 't': 0.27, 'b': 0.02, 'd': 0.02, 'h': 0.02, 'k': 0.02, 'm': 0.02, 'n': 0.02, 'r': 0.02, 's': 0.02, 'v': 0.02, 'w': 0.02, 'y': 0.02}
    sequence = ''
    for i in range(n):
        random_number = lcg() % 1000001
        probability_sum = sum([probabilities[nucleotide]] for nucleotide in probabilities)
        index = int((random_number / probability_sum)) * len(probabilities)
        sequence += list(probabilities.keys()))[index]
    return sequence
def random_homo_sapiens(n):
    probabilities = {'a': 0.302954, 'c': 0.19788, 'g': 0.19764, 't': 0.30149, 'b': 0.02, 'd': 0.02, 'h': 0.02, 'k': 0.02, 'm': 0.02, 'n': 0.02, 'r': 0.02, 's': 0.02, 'v': 0.02, 'w': 0.02, 'y': 0.02}
    sequence = ''
    for i in range(n):
        random_number = lcg() % 1000001
        probability_sum = sum([probabilities[nucleotide]] for nucleotide in probabilities)
        index = int((random_number / probability_sum)) * len(probabilities)
        sequence += list(probabilities.keys()))[index]
    return sequence
def generate_fasta(n):
    alu_sequence = 'GGGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA TCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACT AAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAG GCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCG CCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA'
    iub_sequence = random_iub(n)
    homo_sapiens_sequence = random_homo_sapiens(n))
    print('>ONE Homo sapiens alu\n' + alu_sequence)
    print('>TWO IUB ambiguity codes\n' + iub_sequence)
    print('>THREE Homo sapiens frequency\n' + homo_sapiens_sequence)
if __name__ == '__main__':
    N = int(sys.argv[1])
    generate_fasta(N)
